Introducing ztunnel blog (#12806)
istio/istio.io: content/en/blog/2023/rust-based-ztunnel/index.md

---
title: "Introducing Rust-Based Ztunnel for Istio Ambient Service Mesh"
description: A purpose-built per-node proxy for Istio ambient mesh.

> publishdate: 2023-01-30

Reviewer: What is the target date to publish?
```suggestion
publishdate: 2023-02-28
```

attribution: "Lin Sun (Solo.io), John Howard (Google)"
keywords: [istio,ambient,ztunnel]
---

> The ztunnel (<strong>z</strong>ero <strong>t</strong>rust t<strong>unnel</strong>) component is a purpose-built per-node proxy for Istio ambient mesh. It is responsible for securely connecting and authenticating workloads within ambient mesh. Ztunnel is designed to focus on a small set of features for your workloads in ambient mesh such as mTLS, authentication, L4 authorization and telemetry, without terminating workload HTTP traffic or parsing workload HTTP headers. The ztunnel ensures traffic is efficiently and securely transported to the waypoint proxies, where the full suite of Istio's functionality, such as HTTP telemetry and load balancing, is implemented.

Author: ```suggestion
The ztunnel (zero trust tunnel) component is a purpose-built per-node proxy for Istio ambient mesh. It is responsible for securely connecting and authenticating workloads within ambient mesh. Ztunnel is designed to focus on a small set of features for your workloads in ambient mesh such as mTLS, authentication, L4 authorization and telemetry, without terminating workload HTTP traffic or parsing workload HTTP headers. The ztunnel ensures traffic is efficiently and securely transported to the waypoint proxies, where the full suite of Istio's functionality, such as HTTP telemetry and load balancing, is implemented.
```

> The ztunnel (<strong>z</strong>ero <strong>t</strong>rust t<strong>unnel</strong>) component is a purpose-built per-node proxy for Istio ambient mesh. It is responsible for securely connecting and authenticating workloads within ambient mesh. Ztunnel is designed to be focusing on a small set of features for your workloads in ambient mesh such as mTLS, authentication, L4 authorization and telemetry, without terminating workload HTTP traffic or parsing workload HTTP headers. The ztunnel ensures traffic is efficiently and securely transported to the waypoint proxies, where the full suite of Istio's functionality, such as HTTP telemetry and load balancing, is implemented.

Author: @frankbu @ericvn @howardjohn is there a better way to indicate zero trust tunnel with bold first z and first z and unnel?

Author: This current way fails our lint :-(

Author: removed `<strong>z</strong>ero <strong>t</strong>rust t<strong>unnel</strong>` to pass lint.  If someone has a better idea, happy to restore it!

> Because ztunnel is designed to run on all of your Kubernetes worker nodes, it is critical to keep its resource footprint small. ztunnel is designed to be an invisible (or "ambient") part of your service mesh with minimal impact to your workloads.

Reviewer: ```suggestion
Because ztunnel is designed to run on all of your Kubernetes worker nodes, it is critical to keep its resource footprint small. Ztunnel is designed to be an invisible (or "ambient") part of your service mesh with minimal impact to your workloads.
```

## Ztunnel architecture

Similar to sidecars, ztunnel also serves as an xDS client and CA client:

1. During startup, it securely connects to the Istiod control plane using its
service account token. Once the connection from ztunnel to Istiod is established
securely using TLS, it starts to fetch xDS configuration as an xDS client. This
works similarly to sidecars or gateways or waypoint proxies, except that Istiod
recognizes the request from ztunnel and sends the purpose-built xDS configuration
for ztunnel, which you will learn more about soon.
1. It also serves as a CA client to manage and provision mTLS certificates on behalf of all co-located workloads it manages.
1. As traffic comes in or goes out, it serves as a core proxy that handles the inbound and outbound traffic (either out-of-mesh plain text or in-mesh HBONE) for all co-located workloads it manages.
1. It provides L4 telemetry (metrics and logs) along with an admin server with debugging information to help you debug ztunnel if needed.

{{< image width="100%"
    link="ztunnel-architecture.png"
    caption="Ztunnel architecture"
    >}}

## Why not reuse Envoy?

> When Istio ambient service mesh was announced on Sept 7, 2022, the ztunnel was implemented using Envoy proxy. Given that we use Envoy for the rest of Istio - sidecars, gateways, and waypoint proxies - it was natural for us to start implementing ztunnel using Envoy.

Reviewer: ```suggestion
When Istio ambient service mesh was announced on Sept 7, 2022, the ztunnel was implemented using an Envoy proxy. Given that we use Envoy for the rest of Istio - sidecars, gateways, and waypoint proxies - it was natural for us to start implementing ztunnel using Envoy.
```

## A purpose-built ztunnel

After having trouble bending Envoy to our needs, we started investigating making a purpose-built implementation of the ztunnel. Our hypothesis was that by designing with a single focused use case in mind from the beginning, we could develop a solution that was simpler and more performant than molding a general purpose project to our bespoke use cases. The explicit decision to make ztunnel simple was key to this hypothesis; similar logic wouldn't hold up to rewriting the gateway, for example, which has a huge list of supported features and integrations.

This purpose-built ztunnel involved two key areas:

* The configuration protocol between ztunnel and its Istiod
* The runtime implementation of ztunnel

### Configuration protocol

> Envoy proxies use the [xDS Protocol for configuration](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol). This is a key part of what makes Istio work well, offering rich and dynamic configuration updates. However, as we tread of the beaten path, the config becomes more and more bespoke, which means it's much larger and more expensive to generate. In a sidecar, a single Service with 1 pod, generates roughly ~350 lines of xDS (in YAML), which already has been challenging to scale. The Envoy-based ztunnel was far worse, and in some areas had N^2 scaling attributes.

Reviewer: ```suggestion
Envoy proxies use the [xDS Protocol for configuration](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol). This is a key part of what makes Istio work well, offering rich and dynamic configuration updates. However, as we tread off the beaten path, the config becomes more and more bespoke, which means it's much larger and more expensive to generate. In a sidecar, a single Service with 1 pod, generates roughly ~350 lines of xDS (in YAML), which already has been challenging to scale. The Envoy-based ztunnel was far worse, and in some areas had N^2 scaling attributes.
```

> To keep ztunnel configuration as small as possible, we investigated using a purpose built configuration protocol, that contains precisely the information we need (and nothing more), in an efficient format. For example, a single pod could be represented concisely:

Reviewer: ```suggestion
To keep the ztunnel configuration as small as possible, we investigated using a purpose built configuration protocol, that contains precisely the information we need (and nothing more), in an efficient format. For example, a single pod could be represented concisely:
```

{{< text plaintext >}}
name: helloworld-v1-55446d46d8-ntdbk
namespace: default
serviceAccount: helloworld
node: ambient-worker2
protocol: TCP
status: Healthy
waypointAddresses: []
workloadIp: 10.244.2.8
canonicalName: helloworld
canonicalRevision: v1
workloadName: helloworld-v1
workloadType: deployment
{{< /text >}}

> This information is transported over the xDS transport API, but uses a custom ambient-specific type. Refer to the workload xDS configuration section to learn more about the configuration details.

Author: fixed! thx for the reminder

Before your workloads are included in your ambient mesh, you will still be able to see them in ztunnel’s config dump, as ztunnel is aware of all of the workloads regardless of whether they are ambient enabled or not. For example, below contains a sample workload configuration for a newly deployed helloworld v1 pod which is out-of-mesh indicated by `protocol: TCP`:

{{< text plaintext >}}
{
  "workloads": {
    "10.244.2.8": {
      "workloadIp": "10.244.2.8",
      "protocol": "TCP",
      "name": "helloworld-v1-cross-node-55446d46d8-ntdbk",
      "namespace": "default",
      "serviceAccount": "helloworld",
      "workloadName": "helloworld-v1-cross-node",
      "workloadType": "deployment",
      "canonicalName": "helloworld",
      "canonicalRevision": "v1",
      "node": "ambient-worker2",
      "authorizationPolicies": [],
      "status": "Healthy"
    }
  }
}
{{< /text >}}

After the pod is included in ambient (by labeling the namespace default with `istio.io/dataplane-mode=ambient`), the `protocol` value is replaced with `HBONE`, instructing ztunnel to upgrade all incoming and outgoing communications from the helloworld-v1 pod to be HBONE.

{{< text plaintext >}}

> {
>     "workloadIp": "10.244.2.8",
>     …
>     "protocol": "HBONE",
> }

Reviewer: ```suggestion
{
  "workloads": {
    "10.244.2.8": {
      "workloadIp": "10.244.2.8",
      "protocol": "HBONE",
      ...
}
```

{{< /text >}}

After you deploy any workload level authorization policy, the policy configuration will be pushed as xDS configuration from Istiod to ztunnel and shown under `policies`:

> {{< text plaintext >}}
> "policies": {
>     "default/hw-viewer": {
>       "name": "hw-viewer",
>       "namespace": "default",
>       "scope": "WorkloadSelector",
>       "action": "Allow",
>       "groups": [[[{
>         "principals": [{"Exact": "cluster.local/ns/default/sa/sleep"}]
>       }]]]
>     }
>   }
> …
> {{< /text >}}

Reviewer: ```suggestion
{{< text plaintext >}}
{
  "policies": {
    "default/hw-viewer": {
      "name": "hw-viewer",
      "namespace": "default",
      "scope": "WorkloadSelector",
      "action": "Allow",
      "groups": [[[{
        "principals": [{"Exact": "cluster.local/ns/default/sa/sleep"}]
      }]]]
    }
  }
  ...
}
{{< /text >}}
```

You’ll also notice the workload's configuration is updated with reference to the authorization policy.

{{< text plaintext >}}

> "workloads": {
>     "10.244.2.8": {
>     "workloadIp": "10.244.2.8",
>     …
>     "authorizationPolicies": [
>         "default/hw-viewer"
>     ],
> },

Reviewer: ```suggestion
  "workloads": {
    "10.244.2.8": {
    "workloadIp": "10.244.2.8",
    ...
    "authorizationPolicies": [
        "default/hw-viewer"
    ],
  },
```

{{< text plaintext >}}
2023-02-15T20:40:48.628251Z  INFO inbound{id=4399fa68cf25b8ebccd472d320ba733f peer_ip=10.244.2.5 peer_id=spiffe://cluster.local/ns/default/sa/sleep}: ztunnel::proxy::inbound: got CONNECT request to 10.244.2.8:5000
{{< /text >}}

You can view L4 metrics of your workloads by accessing the `localhost:15020/metrics` API which provides the full set of TCP [standard metrics](/docs/reference/config/metrics/), with same labels that sidecars expose. For example:

{{< text plaintext >}}
istio_tcp_connections_opened_total{
  reporter="source",
  source_workload="sleep",
  source_workload_namespace="default",
  source_principal="spiffe://cluster.local/ns/default/sa/sleep",
  destination_workload="helloworld-v1",
  destination_workload_namespace="default",
  destination_principal="spiffe://cluster.local/ns/default/sa/helloworld",
  request_protocol="tcp",
  connection_security_policy="mutual_tls"
  ...
} 1
{{< /text >}}

If you install Prometheus and Kiali, you can view these metrics easily from Kiali’s UI.

{{< image width="100%"
    link="kiali-ambient.png"
    caption="Kiali dashboard - L4 telemetry provided by ztunnel"
    >}}

## Wrapping up

> We are super excited that the new [Rust-based ztunnel](https://github.com/istio/ztunnel/) is drastically simplified, more lightweight and performant than the prior Envoy-based ztunnel. With the purposefully designed workload xDS for the Rust-based ztunnel, you’ll not only be able to understand the xDS configuration much easily, but also have drastically reduced network traffic and cost between the Istiod control plane and ztunnels. With Istio ambient is merged to upstream master, you can try the new Rust-based ztunnel by grabbing one of the [builds](https://github.com/istio/istio/wiki/Dev-Builds) before it is available in Istio 1.18, or [build](https://github.com/istio/istio/wiki/Dev-Builds) the ztunnel yourself.

Reviewer: ```suggestion
We are super excited that the new [Rust-based ztunnel](https://github.com/istio/ztunnel/) is drastically simplified, more lightweight and performant than the prior Envoy-based ztunnel. With the purposefully designed workload xDS for the Rust-based ztunnel, you’ll not only be able to understand the xDS configuration much more easily, but also have drastically reduced network traffic and cost between the Istiod control plane and ztunnels. With Istio ambient now merged to upstream master, you can try the new Rust-based ztunnel by grabbing one of the [builds](https://github.com/istio/istio/wiki/Dev-Builds) before it is available in Istio 1.18, or [build](https://github.com/istio/istio/wiki/Dev-Builds) the ztunnel yourself.
```
